### PR TITLE
feat(frontend-canister)!: default secure configuration for assets in frontend project template

### PR DESCRIPTION
**Pull request by smallstepman**
# Description

blocked by:
- [x] https://github.com/dfinity/sdk/pull/2489
- [x] #2824 
  
Closes https://dfinity.atlassian.net/browse/SDK-473
Closes https://dfinity.atlassian.net/browse/PSEC-710

# How Has This Been Tested?

e2e

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [ ] I have made corresponding changes to the documentation.

### CODE REVIEW

**Comment by smallstepman** [comment]
@robin-kunzler @eduard-dfinity here is a little script so you can test things on your end.

```shell
git clone https://github.com/dfinity/sdk sdk_sec_headers
cd sdk_sec_headers
git checkout SDK-473-default-security-headers
cargo run -- new sec_headers
cd sec_headers/
cargo run --manifest-path ../Cargo.toml -- start --clean --background
cargo run --manifest-path ../Cargo.toml -- deploy
FE_CAN_ID=$(cargo run --manifest-path=../Cargo.toml -- canister id sec_headers_frontend)
open "http://127.0.0.1:8000/?canisterId=${FE_CAN_ID}"
curl --head "http://127.0.0.1:8000/index.js?canisterId=${FE_CAN_ID}"
curl --head "http://127.0.0.1:8000/main.css?canisterId=${FE_CAN_ID}"
```
### Output
The headers are loaded as expected, and to my untrained eye, everything seems to work as expected also in the browser. 
```
HTTP/1.1 200 OK
content-type: application/javascript
ic-certificate: ic-certificate: certificate=:.....:, tree=:....:
content-security-policy: default-src 'self';script-src 'self' 'unsafe-eval';connect-src 'self' https://ic0.app https://*.ic0.app;img-src 'self' data:;style-src * 'unsafe-inline';style-src-elem * 'unsafe-inline';font-src *;object-src 'none';base-uri 'self';frame-ancestors 'none';form-action 'self';upgrade-insecure-requests;
x-content-type-options: nosniff
x-frame-options: DENY
referrer-policy: same-origin
strict-transport-security: max-age=31536000; includeSubDomains
x-xss-protection: 1; mode=block
permissions-policy: accelerometer=(), ambient-light-sensor=(), autoplay=(), battery=(), camera=(), cross-origin-isolated=(), display-capture=(), document-domain=(), encrypted-media=(), execution-while-not-rendered=(), execution-while-out-of-viewport=(), fullscreen=(), geolocation=(), gyroscope=(), keyboard-map=(), magnetometer=(), microphone=(), midi=(), navigation-override=(), payment=(), picture-in-picture=(), publickey-credentials-get=(), screen-wake-lock=(), sync-xhr=(), usb=(), web-share=(), xr-spatial-tracking=(), clipboard-read=(), clipboard-write=(), gamepad=(), speaker-selection=(), conversion-measurement=(), focus-without-user-activation=(), hid=(), idle-detection=(), interest-cohort=(), serial=(), sync-script=(), trust-token-redemption=(), window-placement=(), vertical-scroll=()
etag: "................"
content-length: 628071
date: Mon, 08 Aug 2022 11:21:12 GMT

HTTP/1.1 200 OK
content-type: text/css
ic-certificate: certificate=:.....:, tree=:....:
content-security-policy: default-src 'self';script-src 'self' 'unsafe-eval';connect-src 'self' https://ic0.app https://*.ic0.app;img-src 'self' data:;style-src * 'unsafe-inline';style-src-elem * 'unsafe-inline';font-src *;object-src 'none';base-uri 'self';frame-ancestors 'none';form-action 'self';upgrade-insecure-requests;
referrer-policy: same-origin
x-xss-protection: 1; mode=block
x-frame-options: DENY
x-content-type-options: nosniff
permissions-policy: accelerometer=(), ambient-light-sensor=(), autoplay=(), battery=(), camera=(), cross-origin-isolated=(), display-capture=(), document-domain=(), encrypted-media=(), execution-while-not-rendered=(), execution-while-out-of-viewport=(), fullscreen=(), geolocation=(), gyroscope=(), keyboard-map=(), magnetometer=(), microphone=(), midi=(), navigation-override=(), payment=(), picture-in-picture=(), publickey-credentials-get=(), screen-wake-lock=(), sync-xhr=(), usb=(), web-share=(), xr-spatial-tracking=(), clipboard-read=(), clipboard-write=(), gamepad=(), speaker-selection=(), conversion-measurement=(), focus-without-user-activation=(), hid=(), idle-detection=(), interest-cohort=(), serial=(), sync-script=(), trust-token-redemption=(), window-placement=(), vertical-scroll=()
strict-transport-security: max-age=31536000; includeSubDomains
etag: "................"
content-length: 537
date: Mon, 08 Aug 2022 11:21:12 GMT

```

**Comment by eduard-dfinity** [comment]
Hey @smallstepman sorry for the late reply. I've reviewed the PR, everything looks good to me: headers are correct and I haven't seen any CSP violation or other errors in the browser. 

btw, it would be nice to have a FOLLOW ticket for the http security headers comments so we can track its progress, would you mind creating it in your side?

**Comment by smallstepman** [comment]
blocked by https://github.com/dfinity/sdk/pull/2489

File: src/dfx/assets/new_project_node_files/webpack.config.js [hunk 1]
```diff
@@ -93,6 +94,11 @@ module.exports = {
       Buffer: [require.resolve("buffer/"), "Buffer"],
       process: require.resolve("process/browser"),
     }),
+    new CopyPlugin({
```
Comment (ericswanson-dfinity) [inline src/dfx/assets/new_project_node_files/webpack.config.js:97]: Why is this needed?  Asset sync copies directly from the `source` directories, so it should see these files without having to copy them.

Comment (smallstepman) [reply src/dfx/assets/new_project_node_files/webpack.config.js:97]: `source` directories in `dfx.json` only include `src/__project_name__/assets` and `src/__project_name__/dist`. Webpack builds stuff from `src/__project_name__/src` into `src/__project_name__/dist`, however since `.ic-assets.json` is not being used in JS files, it's not being moved over from `src` to `dist`, hence we need to manually copy it

Comment (smallstepman) [reply src/dfx/assets/new_project_node_files/webpack.config.js:97]: tho admittedly, that's not the perfect place to bring this change. sorry for that

File: src/dfx/assets/new_project_node_files/src/__project_name___frontend/assets/.ic-assets.json5 [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            // Security: Prevents the browser from interpreting files as a different MIME type to what is specified in the Content-Type header.
+            // See: https://owasp.org/www-project-secure-headers/#x-content-type-options
+            "X-Content-Type-Options": "nosniff",
+
+            // Security: Enables browser features to mitigate some of the XSS attacks. Note that it has to be in mode=block.
+            // See: https://owasp.org/www-community/attacks/xss/
+            "X-XSS-Protection": "1; mode=block"
+        },
+        // redirect all requests from .raw.ic0.app to .ic0.app
+        "allow_raw_access": false
```
Comment (ericswanson-dfinity) [inline src/dfx/assets/new_project_node_files/src/__project_name___frontend/assets/.ic-assets.json5:54]: This is the default, yes?

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,11 @@
 
 ## DFX
 
+### feat(frontend-canister)!: default secure configuration for assets when creating a new frontend project 
+
+- Secure HTTP headers, preventing several typical security vulnerabilities (e.g. XSS, clickjacking, and many more). For more details, see comments in `headers` section in [default `.ic-assets.json5`](./src/dfx/assets/new_project_node_files/src/__project_name___frontend/src/.ic-assets.json5). 
+- Redirect all traffic from `.raw.ic0.app` domain to `.ic0.app`.
```
Comment (ericswanson-dfinity) [inline CHANGELOG.md:10]: Since redirecting from .raw.ic0.app to .ic0.app is the default, let's instead call out that we are showing that configuration in the default starter projects, even though it is already the default.

File: src/dfx/assets/new_project_node_files/src/__project_name___frontend/assets/.ic-assets.json5 (hunk 2 above)
Comment (smallstepman) [reply src/dfx/assets/new_project_node_files/src/__project_name___frontend/assets/.ic-assets.json5:54]: yes, but I thought it would be a good idea to have it mentioned explicitly (if someone would be confused about the traffic being redirected, this would quickly explain things without the need of scrolling through the changelog)

File: src/dfx/assets/new_project_node_files/src/__project_name___frontend/assets/.ic-assets.json5 [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+            // Security: Prevents the browser from interpreting files as a different MIME type to what is specified in the Content-Type header.
+            // See: https://owasp.org/www-project-secure-headers/#x-content-type-options
+            "X-Content-Type-Options": "nosniff",
+
+            // Security: Enables browser features to mitigate some of the XSS attacks. Note that it has to be in mode=block.
+            // See: https://owasp.org/www-community/attacks/xss/
+            "X-XSS-Protection": "1; mode=block"
+        },
+        // redirect all requests from .raw.ic0.app to .ic0.app
```
Comment (ericswanson-dfinity) [inline src/dfx/assets/new_project_node_files/src/__project_name___frontend/assets/.ic-assets.json5:53]:
```suggestion
        // redirect all requests from .raw.ic0.app to .ic0.app (this is the default)
```

**Comment by smallstepman** [comment]
@ericswanson-dfinity please suggest a better PR name
> secure configuration for assets in default frontend canister starter project
?

**Comment by ericswanson-dfinity** [comment]
> @ericswanson-dfinity please suggest a better PR name
> 
> > secure configuration for assets in default frontend canister starter project
> 
> ?
> 
> (I'll modify the changelog header too)

What you put in the changelog looks good
> default secure configuration for assets when creating a new frontend project
or
default secure configuration for assets in frontend project template

**Comment by smallstepman** [comment]
I like the second one better

**Comment by smallstepman** [comment]
ready for re-review @ericswanson-dfinity

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,11 @@
 
 ## DFX
 
+### feat(frontend-canister)!: default secure configuration for assets in frontend project template
+
+- Secure HTTP headers, preventing several typical security vulnerabilities (e.g. XSS, clickjacking, and many more). For more details, see comments in `headers` section in [default `.ic-assets.json5`](https://raw.githubusercontent.com/dfinity/sdk/master/src/dfx/assets/new_project_node_files/src/__project_name___frontend/src/.ic-assets.json5). 
+- Configures `allow_raw_access` option in starter `.ic-assets.json5` config files, with the value set to its default value (which is `false`). We are showing that configuration in the default starter projects for the sake of easier discoverability, even though it's value is set to the default.
```
Comment (ericswanson-dfinity) [inline CHANGELOG.md:10]:
```suggestion
- Configures `allow_raw_access` option in starter `.ic-assets.json5` config files, with the value set to its default value (which is `false`). We are showing that configuration in the default starter projects for the sake of easier discoverability, even though its value is set to the default.
```

**Review by ericswanson-dfinity — APPROVED** [review]
